python/pyspark/sql/types.py

```diff
@@ -442,7 +442,7 @@ def needConversion(self) -> bool:
 
     def toInternal(self, dt: datetime.timedelta) -> Optional[int]:
         if dt is not None:
-            return (math.floor(dt.total_seconds()) * 1000000) + dt.microseconds
```

Comment:
BTW, do you know why and how it adds one more second though?

Reply:
oh because it's fraction in Python. Okay, so it's a very corner case then :-). LGTM